core/src/com/unciv/models/ruleset/unit/BaseUnit.kt

```diff
@@ -175,6 +176,9 @@ class BaseUnit : INamed, IConstruction, CivilopediaText() {
     }
 
     override fun canBePurchased() = "Cannot be purchased" !in uniques
+    override fun canBePurchasedWithFaith(cityInfo: CityInfo) =
```

Comment:
Hm. Maybe IConstruction should be an abstract class instead. I don't like how so many of the functions are identical.

Reply:
So I spend some time trying this, and it turns out this is not a trivial task. This is as `Building()` inherits both from `NamedStats()` and `IConstruction()`, which is not allowed as "only one class may appear in a supertype list". `BaseUnit` has a similar problem as it also inherits from `CivilopediaText`. I'll try to find a workaround tomorrow, otherwise we will just have to deal with the large amount of code duplication between these two classes

Reply:
`CivilopediaText` is just convenience, you can always replace it with `ICivilopediaText` and add the one val manually...

Reply:
Besides, isn't `IConstruction` already `abstract`? And if you want functions common to Building and BaseUnit, functions with implementations are allowed in interfaces, and can be left alone, reused or replaced in subclasses. You want `fun canBePurchased() = "Cannot be purchased" !in uniques` in both classes but not in `PerpetualConstruction`? No problem - intermediate interface.

<details><summary>...</summary>

```kotlin
interface IConstruction : INamed {
    fun getProductionCost(civInfo: CivilizationInfo): Int
    fun getGoldCost(civInfo: CivilizationInfo): Int
    fun isBuildable(cityConstructions: CityConstructions): Boolean
    fun shouldBeDisplayed(cityConstructions: CityConstructions): Boolean
    fun postBuildEvent(cityConstructions: CityConstructions, wasBought: Boolean = false): Boolean  // Yes I'm hilarious.
    fun getResourceRequirements(): HashMap<String,Int>
    fun canBePurchased(): Boolean
}

interface INonPerpetualConstruction: IConstruction {
    override fun canBePurchased(): Boolean = "Cannot be purchased" !in uniques

    var uniques: HashSet<String>
}

open class PerpetualConstruction(override var name: String, val description: String) : IConstruction {
//(...)

class BaseUnit : INamed, INonPerpetualConstruction, CivilopediaText() {
//(... drop canBePurchased in this class and make uniques an override ...)

//etc
```
</details>
... actually runs.